cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/ControlFlowSensitiveDFGPass.kt

```diff
@@ -79,391 +91,238 @@ open class ControlFlowSensitiveDFGPass(ctx: TranslationContext) : TranslationUni
+private fun isSimpleAssignment(currentNode: Node) =
+    currentNode is BinaryOperator &&
+        currentNode.operatorCode == "=" &&
+        (currentNode.lhs as? DeclaredReferenceExpression)?.refersTo != null
+
+/** Checks if the node is an increment or decrement operator (e.g. i++, i--, ++i, --i) */
+private fun isIncOrDec(currentNode: Node) =
+    currentNode is UnaryOperator &&
+        (currentNode.operatorCode == "++" || currentNode.operatorCode == "--") &&
+        (currentNode.input as? DeclaredReferenceExpression)?.refersTo != null
 
-                    // We ignore FieldDeclarations because we cannot be sure how interprocedural
-                    // data flows affect the field. Handling them in the state would only blow up
-                    // the number of paths unnecessarily.
-                    if (lastWriteDecl is FieldDeclaration) continue
+/**
+ * Removes the DFG edges for a potential implicit return statement if it is not in
+ * [reachableReturnStatements].
+ */
+private fun removeUnreachableImplicitReturnStatement(
+    node: Node,
+    reachableReturnStatements: Collection<ReturnStatement>
+) {
+    val lastStatement =
+        ((node as? FunctionDeclaration)?.body as? CompoundStatement)?.statements?.lastOrNull()
+    if (
+        lastStatement is ReturnStatement &&
+            lastStatement.isImplicit &&
+            lastStatement !in reachableReturnStatements
+    )
+        lastStatement.removeNextDFG(node)
+}
 
-                    // Will we generate the same "prev DFG" with the item that is already in the
-                    // list?
-                    allWritesMatch =
-                        allWritesMatch && existingPair.second[lastWriteDecl]?.last() == lastWrite
-                    // All last writes which exist in the "existing pair" match but we have new
-                    // declarations in the current one
-                    allExistingWritesMatch =
-                        allExistingWritesMatch &&
-                            (lastWriteDecl !in existingPair.second ||
-                                existingPair.second[lastWriteDecl]?.last() == lastWrite)
-                }
-                // We found a matching pair in the worklist? Done. Otherwise, maybe there's another
-                // pair...
-                if (allWritesMatch) return true
-                // The new state is a superset of the old one? We will add the missing pieces to the
-                // old one.
-                if (allExistingWritesMatch) {
-                    subsets.add(existingPair)
-                }
-            }
-        }
+/**
+ * A state which actually holds a state for all nodes, one only for declarations and one for
+ * ReturnStatements.
+ */
+class DFGPassState<V>(
+    /** A mapping of a [Node] to its [Lattice]. */
+    var generalState: State<Node, V> = State(),
+    /** A mapping of [Declaration] to its [Lattice]. */
+    var declarationsState: State<Node, V> = State(),
```

Comment:
The declaration state seems to contain more than just declarations - is that intended?

Reply:
Yes it is because the "generalState" collects the nodes which will later receive DFG edges whereas the nodes here don't. I documented this but we can also change the name if necessary.